superbit_lensing/shear_profiles/annular_jmac.py

```diff
@@ -141,11 +255,175 @@ def transform_shears(self, outdir, overwrite=False):
+        '''
 
-        """
-        workhorse; this will contain the actual azimuthal binning
-        """
+        truth_file = self.cat_info['truth_file']
+
+        try:
+            truth = Table.read(truth_file)
+            if self.vb is True:
+                print(f'Read in truth file {truth_file}')
+
+        except FileNotFoundError as fnf_err:
+            print(f'truth catalog {truth_file} not found, check name/type?')
+            raise fnf_err
+
+        cluster_gals = truth[truth['obj_class']=='cluster_gal']
+        cluster_redshift = np.mean(cluster_gals['redshift'])
+
+        truth_bg_gals = truth[truth['redshift'] > cluster_redshift]
+
+        truth_bg_matcher = htm.Matcher(16,
+                                        ra = truth_bg_gals['ra'],
+                                        dec = truth_bg_gals['dec']
+                                        )
+
+        ann_file_ind, truth_bg_ind, dist = truth_bg_matcher.match(
+                                            ra = self.ra,
+                                            dec = self.dec,
+                                            maxmatch = 1,
+                                            radius = 1./3600
+                                            )
+
+        print(f'# {len(dist)} of {len(self.ra)} objects matched to truth background galaxies')
+
+        self.gtan = self.gtan[ann_file_ind]
+        self.gcross = self.gcross[ann_file_ind]
+        self.r = self.r[ann_file_ind]
+        gal_redshifts = truth_bg_gals[truth_bg_ind]['redshift']
+
+        return gal_redshifts
+
+
+    def process_nfw(self, gal_redshifts,outdir='.',overwrite=False):
+        '''
+        Subsample theoretical galaxy redshift distribution to match redshift
+        distribution of detected galaxy catalog using a sort of MC rejection
+        sampling algorithm
+
+        Also compute g_tan, g_cross
+        '''
+
+        if self.nfw_info is not None:
+
+            # Resample so redshift distribution matches input galaxies
+            nfw_tab = self._nfw_resample_redshift(gal_redshifts, outdir, overwrite)
+
+            # Calculate tangential and cross shears for nfw
+            self._nfw_transform_shear(nfw_tab)
+
+            nfw_tab.write(os.path.join(outdir,'subsampled_nfw_cat.fits'),format='fits', overwrite=overwrite)
+
+        else:
+            # No NFW file passed, return None
+            nfw_tab = None
+
+        return nfw_tab
+
+
+
+    def _nfw_resample_redshift(self, gal_redshifts,outdir='.',overwrite=False):
+        '''
+        Subsample theoretical galaxy redshift distribution to match redshift
+        distribution of detected galaxy catalog using a sort of MC rejection
+        sampling algorithm
+        '''
+
+        rng = np.random.default_rng()
+
+        nfw_file = self.nfw_info['nfw_file']
+        nfw = Table.read(nfw_file,format='fits')
+
+        # 34,300 galaxies injected into the simulations
+        pseudo_nfw = rng.choice(nfw, size=34300, replace=False)
```

Comment:
Where does this `34300` come from? If this truly is the number of galaxies injected in the sim, then I think we should get this num from the length of the truth table (which is now passed as an arg in this file now!)

Reply:
You are correct that `34300` is the number of galaxies injected into the simulation, which came from the published COSMOS 2015 source density. However, you make an excellent point! I will modify the code to make the `size` argument of `pseudo_nfw` the length of `truth_cat['obj_class'] == 'gal'`!